### PR TITLE
Build against tf2.8

### PR DESCRIPTION
**Pull request by seanpmorgan**
# Description
Start testing TF2.8 Build

Likely depends on:
https://github.com/tensorflow/addons/pull/2515
https://github.com/tensorflow/addons/pull/2634

### CODE REVIEW

**Comment by seanpmorgan** [comment]
Need to remove MeanMetricWrapper as well:
https://github.com/seanpmorgan/addons/pull/16#issuecomment-963998683

**Comment by bhack** [comment]
@howl-anderson Can you check this error in the CRF test https://github.com/tensorflow/addons/runs/4612816509?check_suite_focus=true#step:4:285

**Comment by gabrieldemarmiesse** [comment]
@seanpmorgan I'll apply a backport label to this PR so that we can have logs and understand why the backport bot is not working. The logs for the PR you linked in  #2632 is too old and github has deleted it.  Of course don't merge the PR that will be created if it works if you don't want it backported :)

**Comment by howl-anderson** [comment]
@bhack OK, I will check that.

**Comment by seanpmorgan** [comment]
@howl-anderson any luck? Appears to be an issue with the test case created for CRF:
` ValueError: No loss found. You may have forgotten to provide a loss argument in the compile() method.`

**Comment by howl-anderson** [comment]
@seanpmorgan Sorry for that. I should be able to fix this in the next few days.

**Comment by howl-anderson** [comment]
@seanpmorgan I had finished the bugfix. Please take a look at seanpmorgan/addons#17

**Comment by seanpmorgan** [comment]
> @seanpmorgan Sorry for that. I should be able to fix this in the next few days.

No need to apologize, it's all volunteer work! Thanks for your help!

File: .github/workflows/release.yml [hunk 1]
```diff
@@ -39,8 +39,8 @@ jobs:
     strategy:
       matrix:
         os: ['macos-10.15', 'windows-latest', 'ubuntu-18.04']
-        py-version: ['3.7', '3.8', '3.9']
-        tf-version: ['2.5.1', '2.7.0']
+        py-version: ['3.7', '3.8', '3.9', '3.10']
+        tf-version: ['2.6.2', '2.8.0']
```
Comment (lgeiger) [inline .github/workflows/release.yml:43]: TensorFlow 2.6 doesn't support Python 3.10, so I think we need to ignore this combination from the build matrix.

**Comment by bhack** [comment]
https://developer.apple.com/forums/thread/700083

**Comment by seanpmorgan** [comment]
`tensorflow-macos` 2.8. is out, will work on resolving issues and aiming to publish tomorrow

**Comment by bhack** [comment]
> `tensorflow-macos` 2.8. is out, will work on resolving issues and aiming to publish tomorrow

@seanpmorgan It isn't available for python 3.10 https://pypi.org/project/tensorflow-macos/#files

**Comment by seanpmorgan** [comment]
> > `tensorflow-macos` 2.8. is out, will work on resolving issues and aiming to publish tomorrow
> 
> @seanpmorgan It isn't available for python 3.10 https://pypi.org/project/tensorflow-macos/#files

Yeah, spoke with @yarri-oss and he said to proceed without it. Currently looking into this duplicate failing metric issue. It has to be related to the test case we fixed in this PR I think.

**Comment by bhack** [comment]
@seanpmorgan Yes but you need to remove your 3.10 upgrade commit on that platform

**Comment by seanpmorgan** [comment]
@qlzh727 @yarri-oss we're seeing some very strange behavior while trying to make this build work that looks specific to Keras object registration. Would someone be able to take a quick look  to see if this rings a bell as I'm struggling to find bandwidth to look into this. It seems someone else is noticing a similar thing and deleting directories within Keras:
https://www.reddit.com/r/tensorflow/comments/srv3fe/weird_assertion_error_assertionerror_duplicate/


On 2.8 :
```
#21 [make_wheel 1/5] RUN python configure.py
#21 sha256:703a4a01db01bf4cc3f9c517a0566cbeeb9e28ef65a48d52b7fcc68eaa232365
#21 2.427 Traceback (most recent call last):
#21 2.427   File "configure.py", line 24, in <module>
#21 2.427     import tensorflow as tf
#21 2.427   File "/usr/local/lib/python3.7/dist-packages/tensorflow/__init__.py", line 473, in <module>
#21 2.427     keras._load()
#21 2.427   File "/usr/local/lib/python3.7/dist-packages/tensorflow/python/util/lazy_loader.py", line 41, in _load
#21 2.427     module = importlib.import_module(self.__name__)
#21 2.427   File "/usr/lib/python3.7/importlib/__init__.py", line 127, in import_module
#21 2.427     return _bootstrap._gcd_import(name[level:], package, level)
#21 2.427   File "/usr/local/lib/python3.7/dist-packages/keras/__init__.py", line 25, in <module>
#21 2.427     from keras import models
#21 2.427   File "/usr/local/lib/python3.7/dist-packages/keras/models.py", line 20, in <module>
#21 2.427     from keras import metrics as metrics_module
#21 2.427   File "/usr/local/lib/python3.7/dist-packages/keras/metrics/__init__.py", line 23, in <module>
#21 2.427     from keras.metrics.base_metric import Metric
#21 2.427   File "/usr/local/lib/python3.7/dist-packages/keras/metrics/base_metric.py", line 25, in <module>
#21 2.427     from keras.engine import base_layer
#21 2.427   File "/usr/local/lib/python3.7/dist-packages/keras/engine/base_layer.py", line 43, in <module>
#21 2.427     from keras.mixed_precision import loss_scale_optimizer
#21 2.427   File "/usr/local/lib/python3.7/dist-packages/keras/mixed_precision/loss_scale_optimizer.py", line 20, in <module>
#21 2.427     from keras.optimizer_experimental import optimizer as optimizer_experimental
#21 2.427   File "/usr/local/lib/python3.7/dist-packages/keras/optimizer_experimental/optimizer.py", line 657, in <module>
#21 2.427     min_consumer_version=1)
#21 2.427   File "/usr/local/lib/python3.7/dist-packages/tensorflow/python/saved_model/revived_types.py", line 133, in register_revived_type
#21 2.427     raise AssertionError(f"Duplicate registrations for type '{identifier}'")
#21 2.427 AssertionError: Duplicate registrations for type 'experimentalOptimizer'
```


On 2.6:
```
#22 255.8 ImportError while loading conftest '/addons/tensorflow_addons/conftest.py'.
#22 256.0 tensorflow_addons/__init__.py:21: in <module>
#22 256.0     from tensorflow_addons import activations
#22 256.0 tensorflow_addons/activations/__init__.py:17: in <module>
#22 256.0     from tensorflow_addons.activations.gelu import gelu
#22 256.0 tensorflow_addons/activations/gelu.py:19: in <module>
#22 256.0     from tensorflow_addons.utils.types import TensorLike
#22 256.0 tensorflow_addons/utils/types.py:24: in <module>
#22 256.0     from keras.engine import keras_tensor
#22 256.0 /usr/local/lib/python3.8/dist-packages/keras/__init__.py:25: in <module>
#22 256.0     from keras import models
#22 256.0 /usr/local/lib/python3.8/dist-packages/keras/models.py:20: in <module>
#22 256.0     from keras import metrics as metrics_module
#22 256.0 /usr/local/lib/python3.8/dist-packages/keras/metrics/__init__.py:23: in <module>
#22 256.0     from keras.metrics.base_metric import Metric
#22 256.0 /usr/local/lib/python3.8/dist-packages/keras/metrics/base_metric.py:25: in <module>
#22 256.0     from keras.engine import base_layer
#22 256.0 /usr/local/lib/python3.8/dist-packages/keras/engine/base_layer.py:40: in <module>
#22 256.0     from keras.mixed_precision import loss_scale_optimizer
#22 256.0 /usr/local/lib/python3.8/dist-packages/keras/mixed_precision/loss_scale_optimizer.py:18: in <module>
#22 256.0     from keras import optimizers
#22 256.0 /usr/local/lib/python3.8/dist-packages/keras/optimizers/__init__.py:31: in <module>
#22 256.0     from keras.optimizers.optimizer_v2 import adadelta as adadelta_v2
#22 256.0 /usr/local/lib/python3.8/dist-packages/keras/optimizers/optimizer_v2/adadelta.py:22: in <module>
#22 256.0     from keras.optimizers.optimizer_v2 import optimizer_v2
#22 256.0 /usr/local/lib/python3.8/dist-packages/keras/optimizers/optimizer_v2/optimizer_v2.py:36: in <module>
#22 256.0     keras_optimizers_gauge = tf.__internal__.monitoring.BoolGauge(
#22 256.0 /usr/local/lib/python3.8/dist-packages/tensorflow/python/eager/monitoring.py:360: in __init__
#22 256.0     super(BoolGauge, self).__init__('BoolGauge', _bool_gauge_methods,
#22 256.0 /usr/local/lib/python3.8/dist-packages/tensorflow/python/eager/monitoring.py:135: in __init__
#22 256.0     self._metric = self._metric_methods[self._label_length].create(*args)
#22 256.0 E   tensorflow.python.framework.errors_impl.AlreadyExistsError: Another metric with the same name already exists.
```

**Comment by seanpmorgan** [comment]
cc @mihaimaruseac to see if anything looks familiar. Seeing as this hit on the 2.6 branch I'm assuming its related to the Keras split out. We didn't fail on testing 2.7 though which is very strange.

More googling shows potential mismatches between TF and Keras, etc.

**Comment by qlzh727** [comment]
We have seen 2.6 error before which is due to a conflicit metric creation between different version of tf and keras. As long as  you install tf and keras at same version (both 2.6 or both 2.7 or both 2.8), then you shouldn't see this error.

For 2.8 error, this is very new to me, and let me add @[chenmoneygithub](https://github.com/chenmoneygithub) for this question.

**Comment by qlzh727** [comment]
And also @k-w-w from save model side.

**Comment by qlzh727** [comment]
From our internal code search, I only found one place for register of "experimentalOptimizer". Could u check if you have multiple version of keras installed? eg both keras and keras-nightly?

**Comment by seanpmorgan** [comment]
Thank you very much @qlzh727 should have suspected our build container since this appeared out of no where this past week:
```
docker run -it --rm tensorflow/build:latest-python3.7 python -m
 pip list
Package                 Version
----------------------- --------------------
keras-nightly           2.9.0.dev2022020908
Keras-Preprocessing     1.1.2
```

@angerson is there a reason we're installing keras-nightly into the build containers?

**Comment by angerson** [comment]
@seanpmorgan I think there is a reason, but I don't know it very deeply. Maybe @mihaimaruseac knows?

If you're building specifically against TF 2.8, you could use the 2.8 containers: https://hub.docker.com/r/tensorflow/build/tags

**Comment by bhack** [comment]
https://github.com/tensorflow/build/blob/master/tf_sig_build_dockerfiles/devel.requirements.txt#L26:L32

**Comment by seanpmorgan** [comment]
@angerson agree we should pin our build container, but it appears that this is installed in all versions?
```
$ docker run -it --rm tensorflow/build:2.8-python3.7 python -m pip list

Package                 Version
----------------------- --------------------
keras-nightly           2.8.0.dev2021121008
Keras-Preprocessing     1.1.2
```

As seen in the above errors this results in some pretty cryptic errors. Can we remove side loaded keras installations?

**Comment by bhack** [comment]
@seanpmorgan With python 3.10 we are going to compile from sources some test dependencies. I've tried to upgrade to a minimal version with python 3.10 wheels with https://github.com/tensorflow/addons/pull/2635/commits/59f28868fd8e9cc3f8bf67a7c48fc761dcd12134.
But I've reverted as they are going to request an updated numpy version that is not compatible with the oldest TF versions in our current supported range.

**Comment by bhack** [comment]
> @angerson agree we should pin our build container, but it appears that this is installed in all versions?
I don't think that the new TF build Docker infra has this support, If you check the mentioned lines `keras-nighlty` is hardcoded in the `requirments`.
 
Also when this will be implemented/fixed upstream I don't know what will be the case. Do we need to prepare each TF wheel with the specific TF build image tag? As generally we support the last three TF releases do w e need to use 3 different images?

**Comment by seanpmorgan** [comment]
> @seanpmorgan With python 3.10 we are going to compile from sources some test dependencies. I've tried to upgrade to a minimal version with python 3.10 wheels with [59f2886](https://github.com/tensorflow/addons/commit/59f28868fd8e9cc3f8bf67a7c48fc761dcd12134). But I've reverted it as they are going to request an updated numpy version that is not compatible with the oldest TF versions in our current supported range.

I think it's easier to skip the 3 skimage tests and wait for minimum numpy version to increase on TF.

**Comment by bhack** [comment]
> I think it's easier to skip the 3 skimage tests and wait for minimum numpy version to increase on TF.
Yes but as I've updated and reverted also Pillow for python3.10 If I remember correctly probably we had also some compile issue for Pillow.

**Comment by bhack** [comment]
And also pytest:
https://stackoverflow.com/questions/69528842/an-error-while-trying-to-execute-tests-on-python-3-10-with-pytest

**Comment by bhack** [comment]
I think that with pytest we have done.

**Comment by bhack** [comment]
@seanpmorgan Ok green light here now. 
I don't know if we will remember about the disabled sk* test in the next release, also we have few declared deprecation that we have never removed.

**Review by bhack — DISMISSED** [review]

**Review by lgeiger — APPROVED** [review]
Let's also add a Python 3.10 version classifier:
https://github.com/tensorflow/addons/blob/0587cde08dfcb014516f4be56e38b7646a6eb721/setup.py#L117-L120

It's only metadata, but it still would be good to do for consistency.

**Review by lgeiger — APPROVED** [review]

**Review by bhack — APPROVED** [review]

**Pull request merged**